src/main/java/application/RequestResponseLoggingFilter.java

```diff
@@ -0,0 +1,85 @@
+
+import beans.auth.HqUserDetailsBean;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.springframework.core.Ordered;
+import org.springframework.core.annotation.Order;
+import org.springframework.web.filter.GenericFilterBean;
+import org.springframework.web.util.ContentCachingResponseWrapper;
+import util.FormplayerHttpRequest;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+
+@Order(Ordered.HIGHEST_PRECEDENCE + 1)
+public class RequestResponseLoggingFilter extends GenericFilterBean {
+
+    private final Log log = LogFactory.getLog(RequestResponseLoggingFilter.class);
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain) throws IOException, ServletException {
+        final HttpServletRequest httpRequest = (HttpServletRequest) request;
+        final HttpServletResponse httpResponse = (HttpServletResponse) response;
+
+        ContentCachingResponseWrapper responseWrapper = new ContentCachingResponseWrapper(httpResponse);
+        filterChain.doFilter(request, responseWrapper);
+
+        // Get a copy of what we are returning for logging
+        String requestBody = IOUtils.toString(httpRequest.getReader());
+        String responseBody = new String(responseWrapper.getContentAsByteArray());
+
+        String restoreAs = null;
+        String url = null;
+        String domain = null;
+        String user = null;
+
+        try {
+            restoreAs = new JSONObject(requestBody).getString("restoreAs");
+        } catch (JSONException e) {
+            // Swallow, restoreAs not always provided in request.
+        }
+
+        if (request instanceof FormplayerHttpRequest) {
+            FormplayerHttpRequest formplayerHttpRequest = (FormplayerHttpRequest) request;
+            domain = formplayerHttpRequest.getDomain();
+            url = new String(formplayerHttpRequest.getRequestURL());
+
+            HqUserDetailsBean userDetailsBean = formplayerHttpRequest.getUserDetails();
+            user = (userDetailsBean != null) ? formplayerHttpRequest.getUserDetails().getUsername() : null;
+        }
+
+        JSONObject logLine = (new JSONObject())
+                .put("requestUrl", url)
+                .put("requestBody", requestBody)
+                .put("projectSpace", domain)
+                .put("username", user)
+                .put("restoreAs", restoreAs)
+                .put("date", this.currentTimeISO8601())
+                .put("responseBody", responseBody);
```

Comment:
Note, you'll find that some of the sample log lines do not have an entry for fields like `username` and `restoreAs`. If these are null it doesn't add it. It looks like we can force a [null value](https://stackoverflow.com/questions/44631604/unable-to-put-null-values-in-json-object) if we want. Also, in my local setup there are always 2 calls for each request, the first has empty fields. Is that normal?

Reply:
restoreAs won't be present in every request. I think username is always populated, because we had some bugs for a while with HMAC auth when it wasn't being sent. Do you have examples of empty username logs? (In theory the server should be ready to accept the request and log the lack of username, even if such requests are malformed, though)

The doublesend is your browser is sending HTTP OPTIONS requests to the server because it's performing a CORS request since you don't have a proxy that's making requests to your local HQ and your formplayer instance look like they are the same terminator.

Reply:
Ahh, then I have not seen an empty username. But that reminds me, I should check what happens when an error is thrown. My intuition is it will return an error response right away and we won't log the request/response, is that behavior acceptable? 

Also see that we have filter tests, I'll add some for this.

Reply:
It's fine for anything that does not result in data access to not be logged in this layer. It will be sufficient for us to have the raw request logs from the gateway there, since it'll provide the same context formplayer had (someone tried to make a request, we didn't let them).